src/cluster_legacy.c

```diff
@@ -3408,12 +3424,17 @@ void clusterLinkConnectHandler(connection *conn) {
          * replaced by the clusterSendPing() call. */
         node->ping_sent = old_ping_sent;
     }
-    /* We can clear the flag after the first packet is sent.
-     * If we'll never receive a PONG, we'll never send new packets
-     * to this node. Instead after the PONG is received and we
-     * are no longer in meet/handshake status, we want to send
-     * normal PING packets. */
-    node->flags &= ~CLUSTER_NODE_MEET;
+    /* NOTE: Assume the current node is A and is asked to MEET another node B.
+     * Once A sends MEET to B, it cannot clear the MEET flag for B until it
+     * gets a response from B. If the MEET packet is not accepted by B due to
+     * link failure, A must continue sending MEET. If A doesn't continue sending
+     * MEET, A will know about B, but B will never add A. Every node always
+     * responds to PINGs from unknown nodes with a PONG, so A will know about B
+     * and continue sending PINGs. But B won't add A until it sees a MEET (or it
+     * gets to know about A from a trusted third node C). In this case, clearing
+     * the MEET flag here leads to asymmetry in the cluster membership. So, we
+     * clear the MEET flag in clusterProcessPacket.
+     */
```

Comment:
nit - I would suggest pulling the conclusion up, which is the last sentence in the comment block below and then explaining the rationale.
```suggestion
    /* NOTE: We clear the `CLUSTER_NODE_MEET` flag in clusterProcessPacket. 
     * 
     * Assume the current node is A and is asked to MEET another node B.
     * Once A sends MEET to B, it cannot clear the MEET flag for B until it
     * gets a response from B. If the MEET packet is not accepted by B due to
     * link failure, A must continue sending MEET. If A doesn't continue sending
     * MEET, A will know about B, but B will never add A. Every node always
     * responds to PINGs from unknown nodes with a PONG, so A will know about B
     * and continue sending PINGs. But B won't add A until it sees a MEET (or it
     * gets to know about A from a trusted third node C). In this case, clearing
     * the MEET flag here leads to asymmetry in the cluster membership. */
```